// @codingame/monaco-vscode-api/vscode/src/vs/workbench/browser/parts/editor/breadcrumbs.js

import { Emitter } from '../../../../base/common/event.js';
import { localize } from '../../../../nls.js';
import { Extensions, ConfigurationScope } from '../../../../platform/configuration/common/configurationRegistry.js';
import '../../../../platform/instantiation/common/extensions.js';
import { Registry } from '../../../../platform/registry/common/platform.js';
import '../../../../platform/instantiation/common/instantiation.js';

class BreadcrumbsService {
    constructor() {
        this._map = ( new Map());
    }
    register(group, widget) {
        if (( this._map.has(group))) {
            throw ( new Error(`group (${group}) has already a widget`));
        }
        this._map.set(group, widget);
        return {
            dispose: () => this._map.delete(group)
        };
    }
    getWidget(group) {
        return this._map.get(group);
    }
}
class BreadcrumbsConfig {
    constructor() {
    }
    static { this.IsEnabled = BreadcrumbsConfig._stub('breadcrumbs.enabled'); }
    static { this.UseQuickPick = BreadcrumbsConfig._stub('breadcrumbs.useQuickPick'); }
    static { this.FilePath = BreadcrumbsConfig._stub('breadcrumbs.filePath'); }
    static { this.SymbolPath = BreadcrumbsConfig._stub('breadcrumbs.symbolPath'); }
    static { this.SymbolSortOrder = BreadcrumbsConfig._stub('breadcrumbs.symbolSortOrder'); }
    static { this.Icons = BreadcrumbsConfig._stub('breadcrumbs.icons'); }
    static { this.TitleScrollbarSizing = BreadcrumbsConfig._stub('workbench.editor.titleScrollbarSizing'); }
    static { this.TitleScrollbarVisibility = BreadcrumbsConfig._stub('workbench.editor.titleScrollbarVisibility'); }
    static { this.FileExcludes = BreadcrumbsConfig._stub('files.exclude'); }
    static _stub(name) {
        return {
            bindTo(service) {
                const onDidChange = ( new Emitter());
                const listener = service.onDidChangeConfiguration(e => {
                    if (e.affectsConfiguration(name)) {
                        onDidChange.fire(undefined);
                    }
                });
                return new (class {
                    constructor() {
                        this.name = name;
                        this.onDidChange = onDidChange.event;
                    }
                    getValue(overrides) {
                        if (overrides) {
                            return service.getValue(name, overrides);
                        }
                        else {
                            return service.getValue(name);
                        }
                    }
                    updateValue(newValue, overrides) {
                        if (overrides) {
                            return service.updateValue(name, newValue, overrides);
                        }
                        else {
                            return service.updateValue(name, newValue);
                        }
                    }
                    dispose() {
                        listener.dispose();
                        onDidChange.dispose();
                    }
                });
            }
        };
    }
}
( Registry.as(Extensions.Configuration)).registerConfiguration({
    id: 'breadcrumbs',
    title: ( localize(3034, "Breadcrumb Navigation")),
    order: 101,
    type: 'object',
    properties: {
        'breadcrumbs.enabled': {
            description: ( localize(3035, "Enable/disable navigation breadcrumbs.")),
            type: 'boolean',
            default: true
        },
        'breadcrumbs.filePath': {
            description: ( localize(
                3036,
                "Controls whether and how file paths are shown in the breadcrumbs view."
            )),
            type: 'string',
            default: 'on',
            enum: ['on', 'off', 'last'],
            enumDescriptions: [
                ( localize(3037, "Show the file path in the breadcrumbs view.")),
                ( localize(3038, "Do not show the file path in the breadcrumbs view.")),
                ( localize(
                3039,
                "Only show the last element of the file path in the breadcrumbs view."
            )),
            ]
        },
        'breadcrumbs.symbolPath': {
            description: ( localize(
                3040,
                "Controls whether and how symbols are shown in the breadcrumbs view."
            )),
            type: 'string',
            default: 'on',
            enum: ['on', 'off', 'last'],
            enumDescriptions: [
                ( localize(3041, "Show all symbols in the breadcrumbs view.")),
                ( localize(3042, "Do not show symbols in the breadcrumbs view.")),
                ( localize(3043, "Only show the current symbol in the breadcrumbs view.")),
            ]
        },
        'breadcrumbs.symbolSortOrder': {
            description: ( localize(3044, "Controls how symbols are sorted in the breadcrumbs outline view.")),
            type: 'string',
            default: 'position',
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            enum: ['position', 'name', 'type'],
            enumDescriptions: [
                ( localize(3045, "Show symbol outline in file position order.")),
                ( localize(3046, "Show symbol outline in alphabetical order.")),
                ( localize(3047, "Show symbol outline in symbol type order.")),
            ]
        },
        'breadcrumbs.icons': {
            description: ( localize(3048, "Render breadcrumb items with icons.")),
            type: 'boolean',
            default: true
        },
        'breadcrumbs.showFiles': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3049, "When enabled breadcrumbs show `file`-symbols."))
        },
        'breadcrumbs.showModules': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3050, "When enabled breadcrumbs show `module`-symbols."))
        },
        'breadcrumbs.showNamespaces': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3051, "When enabled breadcrumbs show `namespace`-symbols."))
        },
        'breadcrumbs.showPackages': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3052, "When enabled breadcrumbs show `package`-symbols."))
        },
        'breadcrumbs.showClasses': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3053, "When enabled breadcrumbs show `class`-symbols."))
        },
        'breadcrumbs.showMethods': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3054, "When enabled breadcrumbs show `method`-symbols."))
        },
        'breadcrumbs.showProperties': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3055, "When enabled breadcrumbs show `property`-symbols."))
        },
        'breadcrumbs.showFields': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3056, "When enabled breadcrumbs show `field`-symbols."))
        },
        'breadcrumbs.showConstructors': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3057, "When enabled breadcrumbs show `constructor`-symbols."))
        },
        'breadcrumbs.showEnums': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3058, "When enabled breadcrumbs show `enum`-symbols."))
        },
        'breadcrumbs.showInterfaces': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3059, "When enabled breadcrumbs show `interface`-symbols."))
        },
        'breadcrumbs.showFunctions': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3060, "When enabled breadcrumbs show `function`-symbols."))
        },
        'breadcrumbs.showVariables': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3061, "When enabled breadcrumbs show `variable`-symbols."))
        },
        'breadcrumbs.showConstants': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3062, "When enabled breadcrumbs show `constant`-symbols."))
        },
        'breadcrumbs.showStrings': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3063, "When enabled breadcrumbs show `string`-symbols."))
        },
        'breadcrumbs.showNumbers': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3064, "When enabled breadcrumbs show `number`-symbols."))
        },
        'breadcrumbs.showBooleans': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3065, "When enabled breadcrumbs show `boolean`-symbols."))
        },
        'breadcrumbs.showArrays': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3066, "When enabled breadcrumbs show `array`-symbols."))
        },
        'breadcrumbs.showObjects': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3067, "When enabled breadcrumbs show `object`-symbols."))
        },
        'breadcrumbs.showKeys': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3068, "When enabled breadcrumbs show `key`-symbols."))
        },
        'breadcrumbs.showNull': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3069, "When enabled breadcrumbs show `null`-symbols."))
        },
        'breadcrumbs.showEnumMembers': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3070, "When enabled breadcrumbs show `enumMember`-symbols."))
        },
        'breadcrumbs.showStructs': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3071, "When enabled breadcrumbs show `struct`-symbols."))
        },
        'breadcrumbs.showEvents': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3072, "When enabled breadcrumbs show `event`-symbols."))
        },
        'breadcrumbs.showOperators': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3073, "When enabled breadcrumbs show `operator`-symbols."))
        },
        'breadcrumbs.showTypeParameters': {
            type: 'boolean',
            default: true,
            scope: ConfigurationScope.LANGUAGE_OVERRIDABLE,
            markdownDescription: ( localize(3074, "When enabled breadcrumbs show `typeParameter`-symbols."))
        }
    }
});

export { BreadcrumbsConfig, BreadcrumbsService };
