// @codingame/monaco-vscode-api/vscode/src/vs/workbench/contrib/terminal/common/terminalStrings.js

import { localize2, localize } from '../../../../nls.js';

const terminalStrings = {
    terminal: ( localize(11882, "Terminal")),
    new: ( localize(11883, "New Terminal")),
    currentSessionCategory: ( localize(11884, 'current session')),
    previousSessionCategory: ( localize(11885, 'previous session')),
    typeTask: ( localize(11886, "Task")),
    typeLocal: ( localize(11887, "Local")),
    actionCategory: ( localize2(11888, "Terminal")),
    focus: ( localize2(11889, "Focus Terminal")),
    kill: {
        ...( localize2(11890, "Kill Terminal")),
        short: ( localize(11891, "Kill")),
    },
    moveToEditor: ( localize2(11892, "Move Terminal into Editor Area")),
    moveIntoNewWindow: ( localize2(11893, "Move Terminal into New Window")),
    newInNewWindow: ( localize2(11894, "New Terminal Window")),
    moveToTerminalPanel: ( localize2(11895, "Move Terminal into Panel")),
    changeIcon: ( localize2(11896, "Change Icon...")),
    changeColor: ( localize2(11897, "Change Color...")),
    split: {
        ...( localize2(11898, "Split Terminal")),
        short: ( localize(11899, "Split")),
    },
    unsplit: ( localize2(11900, "Unsplit Terminal")),
    rename: ( localize2(11901, "Rename...")),
    toggleSizeToContentWidth: ( localize2(11902, "Toggle Size to Content Width")),
    focusHover: ( localize2(11903, "Focus Hover")),
    newWithCwd: ( localize2(11904, "Create New Terminal Starting in a Custom Working Directory")),
    renameWithArgs: ( localize2(11905, "Rename the Currently Active Terminal")),
    scrollToPreviousCommand: ( localize2(11906, "Scroll to Previous Command")),
    scrollToNextCommand: ( localize2(11907, "Scroll to Next Command"))};

export { terminalStrings };
